sos/report/plugins/roce.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, IndependentPlugin
+
+
+class RoCE(Plugin, IndependentPlugin):
+
+    short_desc = 'RoCE (RDMA over Converged Ethernet) information'
+
+    plugin_name = 'roce'
+    profiles = ('hardware',)
+    # rdma installed with iproute2
+    packages = ('iproute2', 'infiniband-diags')
+
+    def setup(self):
+
+        IB_SYS_DIR = "/sys/class/infiniband/"
+        ibs = self.listdir(IB_SYS_DIR) if self.path_isdir(IB_SYS_DIR) else []
+        for ib in ibs:
+
+            # cma_roce_mode and cma_roce_tos will not work for bond devices
+            # ibstat is used to check if the device exist or not, form bond
+            # the command will return an erro:
+            # stat of IB device <device> failed: No such file or directory
+            skip_cma_roce = False
+            ibstat_out = self.exec_cmd(f'ibstat -s {ib}')
+            if ibstat_out['status'] != 0:
+                skip_cma_roce = True
+
+            # dump ECN configuration
+
+            for port in self.listdir(IB_SYS_DIR + ib + "/ports"):
+                # skip Infiniband and IWARP devices
+                try:
+                    p = open(IB_SYS_DIR + ib + "/ports/" + port +
+                             "/link_layer")
+                except IOError:
+                    continue
+                link_layer = p.readline()
+                p.close()
```

Comment:
**Problem**: The file is not properly closed in case of an exception. It is better to use a `with` statement to ensure the file is closed regardless of whether an exception is thrown.

**Suggested Change**: Use a `with` statement to open the file.


**Severity (1 - 4)**: 2 - MINOR

**Lines**: 41-47

```suggestion
                try:
                    with open(IB_SYS_DIR + ib + "/ports/" + port + "/link_layer") as p:
                        link_layer = p.readline()
                except IOError:
                    continue
```

-----

<div align="right">
    <small><i>Generated by <b>Llama 3.1 405b</b></i></small>
</div>
<br />


<div align="right">
  <small>Was this helpful?</small>
  <a href="https://codecritic-feedback.nvidia.com?vote=1&comment_guid=7c5874e1-3bcb-437d-9330-c816b85a06c8&is_github=true" target="_blank">
    <button>👍</button>
  </a>
  <a href="https://codecritic-feedback.nvidia.com?vote=-1&comment_guid=7c5874e1-3bcb-437d-9330-c816b85a06c8&is_github=true" target="_blank">
    <button>👎</button>
  </a>
</div>